### PR TITLE
first assignment: initial commit

### PR DESCRIPTION
**Pull request by yuvdanilova**
## Task 1, assignment1_permutation.py

Given two strings, write a method to decide if one is a permutation of the other.

**Input:** two non-empty strings separated by a space or a new line
> abc cba

**Output:** boolean, True if two lines are permutations of each other, False otherwise

> True

A naive solution is to sort both strings and compare them, which could be done in O(nlogn) time, where n is the maximum length of two strings.

It is not necessary though to compare strings if they are of different length: by definition these can't be permutations of each other, so we check for this case in the very beginning and if it takes place, return False.

Then notice, that anyhow we will have to check all the symbols of both strings, thus time complexity can't be less than O(n). 

Let's create an auxiliary dictionary that would store non-zero symbol counts. For the first string, iterate over its letters and count their occurrences. Then, for the second string, iterate over its symbols, subtract 1 from corresponding entries in the dict. If we get a zero value, delete key-value pair from the dict. If there's no such key in the dict, return False, as we found a letter that is not in the first string. If the resulting dictionary is empty, the strings have the same set of letters, return True.

The advantage of using a dict is that operations we need are O(1), so it's fast. We don't use sets as we need to check multiple occurrences of the same letter, a python list has O(n) for checking if there's such value. 

Seems to be O(n) time, O(n) space.

## Task 2, assignment1_sll.py

Implement an algorithm to find the kth to last element of a singly linked list.

**Input:** elements of the linked list from the last one to the first one separated by space, then required number k

> a b c d
2

This input means that a singly linked list d -> c -> b -> a is created

**Output:** k'th element from the end

> c

Suppose that 0'th element from the end is the last element itself, and the last element is the one that points to None.

Also suppose that we are trying to solve the task for the classic implementation of a SLL, where the total number of entries n is unknown. In this case, it could be solved just by iterating n - k times.

In this task it doesn't seem to be possible to get less than O(n) complexity as the length of the list is unknown, we need to learn it anyway, thus have to iterate over all values.

A naive solution is to count the number of elements in O(n), then solve the task as we know this number. Total time complexity is O(n) + O(n - k) = O(n) which is not so bad itself.

But we can iterate only once over the list: just store the element that is k elements before on each step (on the (k+1)-st step we know, that the head is the solution, on the (k+2)-nd - next of the head and so on). If there are less elements than k + 1, raise an error as there are not enough values to give an answer.

The total time complexity is O(n), storage O(1)

### CODE REVIEW

File: assignment1_permutation.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
+# Uses python3
```
Comment (ariava) [inline assignment1_permutation.py]: You can write:

# /usr/bin/env python3

to force use of the python3 interpreter when running this script.

File: assignment1_permutation.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+        return letters == {}
+
+
+if __name__ == '__main__':
+    input = sys.stdin.read()
+    strings = input.split()
+    if len(strings) != 2:
+        raise RuntimeError('2 strings expected, got {0}'.format(len(strings)))
+    else:
+        print(check_str_permutation(strings[0], strings[1]))
```
Comment (ariava) [inline assignment1_permutation.py]: Usually it's better to automate testing and not require user input.

You can use the "assert" statement to perform automated testing (e.g. see https://www.tutorialspoint.com/python/assertions_in_python.htm). If any of the checks fail, assert will break the program execution with an AssertionError.

File: assignment1_sll.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    *ssl_elements, k = input.split()
+
+    ssl = SinglyLinkedList()
+    for el in ssl_elements:
+        ssl.append(el)
+
+    if not k.isdigit():
+        raise RuntimeError('k is not a digit: {0}'.format(k))
+
+    print(ssl.kth_from_end(int(k)))
```
Comment (ariava) [inline assignment1_sll.py]: Please add automated testing to this one, with few test cases that pass and few that fail (which are expressed with "assert <check>" and "assert not <check>").

**Review by ariava — CHANGES_REQUESTED** [review]
Please add few automated test cases to each of the files.

**Comment by yuvdanilova** [comment]
Added automatic testing using 'unittest' library, which seems to be a nice tool for testing things like throwing errors. Fixed 'Uses python3' issue

**Pull request merged**